tokio/src/sync/mpsc/bounded.rs

```diff
@@ -849,10 +861,67 @@ impl<T> Sender<T> {
     /// }
     /// ```
     pub async fn reserve(&self) -> Result<Permit<'_, T>, SendError<()>> {
-        self.reserve_inner().await?;
+        self.reserve_inner(1).await?;
         Ok(Permit { chan: &self.chan })
     }
 
+    /// Waits for channel capacity. Once capacity to send `n` message is
+    /// available, it is reserved for the caller.
+    ///
+    /// If the channel is full, the function waits for the number of unreceived
+    /// messages to become less than the channel capacity. Capacity to send `n`
+    /// message is reserved for the caller. A [`ManyPermit`] is returned to track
+    /// the reserved capacity. The [`send`] function on [`ManyPermit`] consumes the
+    /// reserved capacity.
+    ///
+    /// Dropping [`ManyPermit`] without sending a message releases the capacity back
+    /// to the channel.
+    ///
+    /// [`ManyPermit`]: ManyPermit
+    /// [`send`]: ManyPermit::send
+    ///
+    /// # Cancel safety
+    ///
+    /// This channel uses a queue to ensure that calls to `send` and `reserve`
+    /// complete in the order they were requested. Cancelling a call to
+    /// `reserve` makes you lose your place in the queue.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = mpsc::channel(2);
+    ///
+    ///     // Reserve capacity
+    ///     let mut permit = tx.reserve_many(2).await.unwrap();
+    ///
+    ///     // Trying to send directly on the `tx` will fail due to no
+    ///     // available capacity.
+    ///     assert!(tx.try_send(123).is_err());
+    ///
+    ///     // Sending on the permit succeeds
+    ///     permit.send(456).unwrap();
+    ///     permit.send(457).unwrap();
+    ///
+    ///     // The third should fail due to no available capacity
+    ///     permit.send(458).unwrap_err();
+    ///     
+    ///     // The value sent on the permit is received
+    ///     assert_eq!(rx.recv().await.unwrap(), 456);
+    ///     assert_eq!(rx.recv().await.unwrap(), 457);
+    /// }
+    /// ```
+    pub async fn reserve_many(&self, n: u32) -> Result<ManyPermit<'_, T>, SendError<()>> {
```

Comment:
Perhaps we could return an iterator of permits? I think that's a bit nicer than the `send` call that just fails once you've used up the permits.

Reply:
That's a good idea! So we could also remove the `ManyPermit` struct and logic and simply return an iterator of `Permit` ?

Reply:
That makes sense to me. We still need to make sure that the destructor of the iterator puts back permits if they haven't been taken from the iterator though.